File: 61.旋转链表.py
#
# @lc app=leetcode.cn id=61 lang=python3
#
# [61] 旋转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if head == None or head.next == None:
            return head
        length = 1
        oldTail = head
        while oldTail.next:
            length += 1
            oldTail = oldTail.next
        k %= length
        oldTail.next = head
        newTail = head
        for i in range(length-k-1):
            newTail = newTail.next
        newHead = newTail.next
        newTail.next = None
        return newHead

# @lc code=end

File: 61.旋转链表.cpp
/*
 * @lc app=leetcode.cn id=61 lang=cpp
 *
 * [61] 旋转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *rotateRight(ListNode *head, int k)
    {
        if (head == NULL || head->next == NULL)
            return head;
        int len = 1;
        ListNode *oldTail = head;
        for (; oldTail->next != NULL; len++)
        {
            oldTail = oldTail->next;
        }
        k %= len;
        oldTail->next = head;
        ListNode *newTail = head;
        for (int i = 0; i < len - k - 1; i++)
        {
            newTail = newTail->next;
        }
        ListNode *newHead = newTail->next;
        newTail->next = NULL;
        return newHead;
    }
};
// @lc code=end
